### DISCUSSION TITLE
Elements in TF Tree not connected

### DISCUSSION THREAD

**Discussion by sebastianfer** [iRobotEducation/create3_docs#118] [Questions] [answered]
I figured out my SLAM toolbox is not working because there are two elements not connected (odom and base_link). Could someone tell me how to fix that in the static transform publisher? I am attaching what the pdf shows me, and I am using iRobot's sensors_launch.py and create3_lidar package: https://github.com/iRobotEducation/create3_examples/tree/e0dcd78f9b00c6fa1533d1d4cb37542c29095edc/create3_lidar

![Screenshot from 2022-07-22 10-47-47](https://user-images.githubusercontent.com/108418842/180469420-b47035d5-bf1a-40a4-9e80-b8e59b73e57e.png)

Could the location of the camera be affecting this? It looks like this: 

![IMG_8018](https://user-images.githubusercontent.com/108418842/180469917-15893a0a-3556-481f-8b42-b6b5541c97f6.jpg)

**Comment by justinIRBT** [comment, marked as answer]
Odom->base_link and odom->base_footprint are published on /tf by the Create3.  Is the computer you are running slam_toolbox on able to see the create topics?

**Reply by sebastianfer** [reply]
Yes, I am able to see the Create topics: 
/Create/battery_state
/Create/cmd_audio
/Create/cmd_lightring
/Create/cmd_vel
/Create/dock
/Create/hazard_detection
/Create/imu
/Create/interface_buttons
/Create/ir_intensity
/Create/ir_opcode
/Create/kidnap_status
/Create/mouse
/Create/odom
/Create/robot_state/transition_event
/Create/slip_status
/Create/static_transform/transition_event
/Create/stop_status
/Create/system_monitor/transition_event
/Create/tf
/Create/tf_static
/Create/wheel_status
/Create/wheel_ticks
/Create/wheel_vels
/map
/map_metadata
/parameter_events
/rosout
/scan
/slam_toolbox/feedback
/slam_toolbox/graph_visualization
/slam_toolbox/scan_visualization
/slam_toolbox/update
/tf
/tf_static


And this is what I get when I run ros2 topic echo /Create/tf : 
transforms:
- header:
   stamp:
     sec: 1658512482
     nanosec: 116973758
   frame_id: odom
 child_frame_id: base_link
 transform:
   translation:
     x: -0.024669216945767403
     y: -0.054172273725271225
     z: -0.00029985440778546035
   rotation:
     x: -0.024451902136206627
     y: -0.02256624586880207
     z: -0.41536492109298706
     w: 0.9090461730957031
- header:
   stamp:
     sec: 1658512482
     nanosec: 116973758
   frame_id: odom
 child_frame_id: base_footprint
 transform:
   translation:
     x: -0.024726106389051594
     y: -0.05417174040327133
     z: 0.0
   rotation:
     x: 0.0
     y: 0.0
     z: -0.41523534059524536
     w: 0.9097140431404114


I can see the transforms in the /tf, which makes me wonder why they don't appear in the frame

**Reply by justinIRBT** [reply]
It looks like you are using a namespace for your Create.  The general tf tree used by the system is looking on `/tf`.  You will need to re-broadcast your `/Create/tf` tree with a transform broadcaster to get the `odom->base_link` and `odom->base_footprint` on `/tf`.  Here is an example of writing a transform broadcaster:
C++
https://docs.ros.org/en/galactic/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Broadcaster-Cpp.html
Python
https://docs.ros.org/en/galactic/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Broadcaster-Py.html

You will need to subscribe to `/Create/tf` pull out the transform and re-broadcast with the Transform broadcaster.  Alternatively, if the namespace isn't necessary in your use case (if you only have one robot on the network), then if you remove the namespace, it should just work.

More explanation for why namespace is used with tf was discussed here:
https://github.com/iRobotEducation/create3_docs/discussions/39

**Reply by sebastianfer** [reply]
Thanks! Now everything's working! I still added nodes for the transforms of map->odom, and odom->base_link. After doing that, I got a full tf tree. I got the values for the arguments from by running ros2 topic echo /tf.

**Reply by jmm-slamcore** [reply]
If we want to use our own odom->base_link transform provided by a SLAM system instead of the Create3 transform, what is the recommended approach, as disabling tf publishing is not currently possible on the create3?

**Reply by shamlian** [reply]
For now, I would recommend giving Create 3 a namespace and generating your own tf tree to suit your needs (see #39 ); we will endeavor to parameterize this (see #144 ) for a future build.
